pkg/metrics_store/metrics_writer.go

```diff
@@ -82,20 +87,59 @@ func (m MetricsWriter) WriteAll(w io.Writer) error {
 	return nil
 }
 
-// SanitizeHeaders removes duplicate headers from the given MetricsWriterList for the same family (generated through CRS).
-// These are expected to be consecutive since G** resolution generates groups of similar metrics with same headers before moving onto the next G** spec in the CRS configuration.
-func SanitizeHeaders(writers MetricsWriterList) MetricsWriterList {
+// SanitizeHeaders sanitizes the headers of the given MetricsWriterList.
+func SanitizeHeaders(contentType string, writers MetricsWriterList) MetricsWriterList {
 	var lastHeader string
 	for _, writer := range writers {
 		if len(writer.stores) > 0 {
 			for i, header := range writer.stores[0].headers {
+
+				// Nothing to sanitize.
+				if len(header) == 0 {
+					continue
+				}
+
+				// Removes duplicate headers from the given MetricsWriterList for the same family (generated through CRS).
+				// These are expected to be consecutive since G** resolution generates groups of similar metrics with same headers before moving onto the next G** spec in the CRS configuration.
+				// Skip this step if we encounter a repeated header, as it will be removed.
+				if header != lastHeader && strings.HasPrefix(header, "# HELP") {
+
+					// If the requested content type was proto-based, replace "info" and "statesets" with "gauge", as they are not recognized by Prometheus' protobuf machinery.
+					if strings.HasPrefix(contentType, expfmt.ProtoType) {
+						infoTypeString := metric.Info.String()
+						stateSetTypeString := metric.StateSet.String()
+						if strings.HasSuffix(header, infoTypeString) {
+							header = header[:len(header)-len(infoTypeString)] + string(metric.Gauge)
+							writer.stores[0].headers[i] = header
+						}
+						if strings.HasSuffix(header, stateSetTypeString) {
+							header = header[:len(header)-len(stateSetTypeString)] + string(metric.Gauge)
+							writer.stores[0].headers[i] = header
+						}
+					}
+				}
+
+				// Nullify duplicate headers after the sanitization to not miss out on any new candidates.
 				if header == lastHeader {
 					writer.stores[0].headers[i] = ""
-				} else {
-					lastHeader = header
 				}
+
+				// Update the last header.
+				lastHeader = header
 			}
 		}
 	}
+
+	// Remove all empty headers.
+	for _, writer := range writers {
+		if len(writer.stores) > 0 {
+			for i := len(writer.stores[0].headers) - 1; i >= 0; i-- {
+				if writer.stores[0].headers[i] == "" {
+					writer.stores[0].headers = append(writer.stores[0].headers[:i], writer.stores[0].headers[i+1:]...)
+				}
+			}
+		}
+	}
```

Comment:
why do we do that? The order of the headers is very important so modifying it doesn't like a good idea

Reply:
`G**` resolutions can [result in duplicated families](https://github.com/rexagod/kube-state-metrics/blob/main/pkg/metrics_store/metrics_writer.go#L95). I noticed that in such cases we modified the headers to an empty string, but since that is not a valid metric byte line, I added some functionality to strip those and enforced that using some [tests here](https://github.com/kubernetes/kube-state-metrics/pull/2270/files#diff-a2fba22b7063f6a5f98f7c677327f9d87bc5f07d84cd567009a1f41ce6799adeR298-R315).

Reply:
That sounds like a bug where we don't account for the version anywhere besides in the `customresource_version` label. So we create identical stores but for different resources no?

Do you mind to take that to a separate PR? It sounds like a different issue than the original one.